// tauri-plugin-alarm-api/dist-js/index.cjs
'use strict';

var core = require('@tauri-apps/api/core');

/**
 * アラームをスケジュールする。
 *
 * Android では AlarmManager を使用してアラームを設定し、
 * 発火時に通知を表示する。
 *
 * @example
 * ```ts
 * await setAlarm({
 *   id: 1,
 *   title: '起床時間です',
 *   message: 'おはようございます！',
 *   triggerAtMs: Date.now() + 8 * 60 * 60 * 1000,  // 8 時間後
 * })
 * ```
 */
async function setAlarm(options) {
    return await core.invoke('plugin:alarm|set_alarm', {
        payload: options,
    });
}
/**
 * スケジュール済みのアラームをキャンセルする。
 *
 * @param id キャンセルするアラームの ID
 */
async function cancelAlarm(id) {
    await core.invoke('plugin:alarm|cancel_alarm', {
        payload: { id },
    });
}
/**
 * 現在スケジュールされているアラームの一覧を取得する。
 */
async function listAlarms() {
    const result = await core.invoke('plugin:alarm|list_alarms', {
        payload: {},
    });
    return result.alarms;
}
/**
 * 正確なアラームをスケジュールする権限があるか確認する。
 *
 * Android 12 (API 31) 以上では SCHEDULE_EXACT_ALARM 権限が必要。
 * Android 11 以下では常に true を返す。
 */
async function checkExactAlarmPermission() {
    return await core.invoke('plugin:alarm|check_exact_alarm_permission', { payload: {} });
}
/**
 * 正確なアラームの権限設定画面を開く（Android 12+ のみ有効）。
 *
 * @example
 * ```ts
 * const { canScheduleExactAlarms } = await checkExactAlarmPermission()
 * if (!canScheduleExactAlarms) {
 *   await openExactAlarmSettings()
 * }
 * ```
 */
async function openExactAlarmSettings() {
    await core.invoke('plugin:alarm|open_exact_alarm_settings', {
        payload: {},
    });
}

exports.cancelAlarm = cancelAlarm;
exports.checkExactAlarmPermission = checkExactAlarmPermission;
exports.listAlarms = listAlarms;
exports.openExactAlarmSettings = openExactAlarmSettings;
exports.setAlarm = setAlarm;
